src/daft-plan/src/physical_ops/shuffle_exchange.rs

```diff
@@ -0,0 +1,151 @@
+};
+
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct ShuffleExchange {
+    pub input: PhysicalPlanRef,
+    pub strategy: ShuffleExchangeStrategy,
+}
+
+impl ShuffleExchange {
+    /// Retrieve the output clustering spec associated with this ShuffleExchange
+    pub fn clustering_spec(&self) -> Arc<ClusteringSpec> {
+        match &self.strategy {
+            ShuffleExchangeStrategy::NaiveFullyMaterializingMapReduce { target_spec } => {
+                target_spec.clone()
+            }
+            ShuffleExchangeStrategy::SplitOrCoalesceToTargetNum {
+                target_num_partitions,
+            } => Arc::new(ClusteringSpec::Unknown(UnknownClusteringConfig::new(
+                *target_num_partitions,
+            ))),
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub enum ShuffleExchangeStrategy {
+    /// Fully materialize the data after the Map, and then pull results from the Reduce.
+    NaiveFullyMaterializingMapReduce { target_spec: Arc<ClusteringSpec> },
+
+    /// Sequentially splits/coalesce partitions in order to meet a target number of partitions
+    SplitOrCoalesceToTargetNum { target_num_partitions: usize },
+}
+
+impl ShuffleExchange {
+    pub fn multiline_display(&self) -> Vec<String> {
+        let mut res = vec![];
+        res.push("ShuffleExchange:".to_string());
+        match &self.strategy {
+            ShuffleExchangeStrategy::NaiveFullyMaterializingMapReduce { target_spec } => {
+                res.push("  Strategy: NaiveFullyMaterializingMapReduce".to_string());
+                res.push(format!("  Target Spec: {:?}", target_spec));
+                res.push(format!(
+                    "  Number of Partitions: {} → {}",
+                    self.input.clustering_spec().num_partitions(),
+                    target_spec.num_partitions(),
+                ));
+            }
+            ShuffleExchangeStrategy::SplitOrCoalesceToTargetNum {
+                target_num_partitions,
+            } => {
+                let input_num_partitions = self.input.clustering_spec().num_partitions();
+                res.push("  Strategy: SplitOrCoalesceToTargetNum".to_string());
+                res.push(format!(
+                    "  {} Partitions: {} → {}",
+                    if input_num_partitions >= *target_num_partitions {
+                        "Coalescing"
+                    } else {
+                        "Splitting"
+                    },
+                    input_num_partitions,
+                    target_num_partitions,
+                ));
+            }
+        }
+        res
+    }
+}
+
+impl_default_tree_display!(ShuffleExchange);
+
+/// Builder of ShuffleExchanges
+///
+/// This provides an abstraction where we can select the most appropriate strategies based on various
+/// heuristics such as number of partitions and the currently targeted backend's available resources.
+pub struct ShuffleExchangeBuilder {
```

Comment:
Good callout. I changed this to be a Factory instead of adding a bunch of `::new` methods directly on ShuffleExchange, because I do foresee that we will want the factory to be configurable.

E.g. the factory should be parametrized by config variables as well as other things such as statistics/estimations on the number of partitions. Then when you create the shuffle exchange, it will dynamically choose the correct ShuffleExchange implementation based on the data available to it.